### PR TITLE
Self referencing scene bug

### LINKED ISSUE #175: Fix self-referencing scenes
Scenes are currently able to reference themselves or scenes higher up in the scene graph.
This creates an infinite loop when parsing the tree.

To reproduce:
Create a scene.
Save it.
Instantiate it in it self.
Change scene.
Try to change back.
Crash.

More info curtesy of Ludvig Arlebrink:
You should not be able to instantiate scene to self (scene 1), meaning, the current scene should not come up as an option in the editor, since this will cause an infinite loop. However, if you instantiate another scene (scene 2) and in scene 2 you instantiate scene 1 the same error, infinite loop, will occur.

### PR DESCRIPTION
**Pull request by 1uk45**
fixes #175 
I've made so that when creating same instance of scene it will be copied into a new file. But you have to save the scene before so that it has a file to read from.

### CODE REVIEW

File: src/Editor/GUI/Editors/SceneEditor.cpp [hunk 1]
```diff
@@ -137,8 +137,13 @@ void SceneEditor::ShowEntity(Entity* entity) {
         ImGui::Separator();
         
         for (const std::string& scene : Hymn().scenes) {
+            bool isSameScene = false;
+            if (Hymn().scenes[sceneIndex] == scene)
+            {
+                isSameScene = true;
+            }
```
Comment (ludvigarlebrink) [inline src/Editor/GUI/Editors/SceneEditor.cpp]: Remove brackets.

File: src/Engine/Entity/Entity.cpp [hunk 2]
```diff
@@ -65,20 +65,43 @@ bool Entity::HasChild(const Entity* check_child, bool deep) const {
...
+            copyOfFile.close();
+            //Reads from copy.
+            std::ifstream readFromCopyFile;
+            readFromCopyFile.open(filenameCopy);
+            readFromCopyFile >> rootCopy;
+            readFromCopyFile.close();
+            //Loads copy of scene.
+            child->Load(rootCopy);
+            child->scene = true;
+            child->sceneName = name + "_copy";
```
Comment (ludvigarlebrink) [inline src/Engine/Entity/Entity.cpp]: Put everything in this if statement into a help function maybe?

File: src/Engine/Entity/Entity.hpp [hunk 3]
```diff
@@ -58,9 +59,10 @@ class Entity {
         /// Instantiate a scene as a child to this entity.
         /**
          * @param name The name of the scene to instantiate.
+         * @param isSameScene set to true if it is the same scene you want to istantiate.
          * @return The created root entity of the scene.
          */
-        Entity* InstantiateScene(const std::string& name);
+        Entity* InstantiateScene(const std::string& name, bool isSameScene);
```
Comment (ludvigarlebrink) [inline src/Engine/Entity/Entity.hpp]: Entity* InstantiateScene(const std::string& name, bool isSameScene = false); maybe?

File: src/Engine/Entity/Entity.hpp [hunk 4]
```diff
@@ -58,9 +59,10 @@ class Entity {
         /// Instantiate a scene as a child to this entity.
         /**
          * @param name The name of the scene to instantiate.
+         * @param isSameScene set to true if it is the same scene you want to istantiate.
```
Comment (ludvigarlebrink) [inline src/Engine/Entity/Entity.hpp]: Capital S on set.

File: src/Engine/Entity/Entity.cpp [hunk 5]
```diff
@@ -65,20 +65,43 @@ bool Entity::HasChild(const Entity* check_child, bool deep) const {
...
+        } else { //dont need to be a copy
+            Json::Value root;
+            std::ifstream file(filename);
+            file >> root;
+            file.close();
+            child->Load(root);
+            child->scene = true;
+            child->sceneName = name;
```
Comment (ludvigarlebrink) [inline src/Engine/Entity/Entity.cpp]: Put everything in this else statement into a help function maybe?

File: src/Engine/Entity/Entity.hpp [hunk 6]
```diff
@@ -1,11 +1,12 @@
 #pragma once
-
 #include <map>
 #include <vector>
 #include <typeinfo>
 #include "../Entity/World.hpp"
 #include <json/json.h>
 #include "../Component/SuperComponent.hpp"
+#include <fstream>
+
```
Comment (ludvigarlebrink) [inline src/Engine/Entity/Entity.hpp]: Remove newline.

File: src/Engine/Entity/Entity.cpp [hunk 7]
```diff
@@ -65,20 +65,43 @@ bool Entity::HasChild(const Entity* check_child, bool deep) const {
...
     if (FileSystem::FileExists(filename.c_str())) {
-        Json::Value root;
-        std::ifstream file(filename);
-        file >> root;
-        file.close();
-        child->Load(root);
-        
-        child->scene = true;
-        child->sceneName = name;
+        //If it needs to be a copy.
```
Comment (ludvigarlebrink) [inline src/Engine/Entity/Entity.cpp]: Space between // and comment. Example: // If it needs to be a copy.

File: src/Engine/Entity/Entity.hpp (hunk 4 above)
Comment (ludvigarlebrink) [inline src/Engine/Entity/Entity.hpp]: Change "isSameScene" name to example: "copy",  would be more descriptive of what it actually does.

**Comment by Chainsawkitten** [comment]
I don't really understand how this solves the issue. The problem is that a loop should not be allowed to exist.

**Comment by 1uk45** [comment]
@Chainsawkitten there is no loop with this solution. With this we can create prefabs with all scenes even if it is the same one.

**Pull request closed without merging**